utils_gorm.go

```diff
@@ -0,0 +1,181 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/iver-wharf/wharf-core/pkg/ginutil"
+	"gorm.io/gorm"
+	"gorm.io/gorm/clause"
+)
+
+func findDBPaginatedSliceAndTotalCount(dbQuery *gorm.DB, limit, offset int, slice interface{}, totalCount *int64) error {
+	err := dbQuery.Scopes(optionalLimitOffsetScope(limit, offset)).Find(slice).Error
+	if err != nil {
+		return err
+	}
+
+	return dbQuery.Count(totalCount).Error
+}
+
+func fetchDatabaseObjByID(c *gin.Context, db *gorm.DB, modelPtr interface{}, id uint, name, whenMsg string) bool {
+	var spacedWhenMsg string
+	if whenMsg != "" {
+		spacedWhenMsg = " " + whenMsg
+	}
+	if err := db.First(modelPtr, id).Error; err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			ginutil.WriteDBNotFound(c, fmt.Sprintf(
+				"%s with ID %d was not found%s.",
+				strings.Title(name), id, spacedWhenMsg))
+		} else {
+			ginutil.WriteDBReadError(c, err, fmt.Sprintf(
+				"Failed fetching %s with ID %d from database%s.",
+				strings.ToLower(name), id, spacedWhenMsg))
+		}
+		return false
+	}
+	return true
+}
+
+func optionalLimitOffsetScope(limit, offset int) func(*gorm.DB) *gorm.DB {
+	if limit <= 0 {
+		return gormIdentityScope
+	}
+	if offset <= 0 {
+		offset = 0
+	}
+	return func(db *gorm.DB) *gorm.DB {
+		return db.Clauses(clause.Limit{
+			Limit:  limit,
+			Offset: offset,
+		})
+	}
+}
+
+func optionalTimeRangeScope(column string, min, max *time.Time) func(*gorm.DB) *gorm.DB {
```

Comment:
Do we want to add a check for `max < min` here, or do we get an error for that when running the query?

Reply:
Good question! As the `x BETWEEN ? AND ?` operator in SQL is defined as syntactic sugar for `x >= ? AND x <= ?` then I assumed it didn't matter.

Apparently Postgres has a `x BETWEEN SYMMETRIC ? AND ?`, which automatically flips the operands for you if `min > max`.

Did some manual testing, and both Sqlite3 and PostgreSQL just returned an empty set, which was what I was hoping for.

And if the database dislikes the input anyway, I don't see any problem with wharf-api responding with a "DB read error" problem. Should maybe be a 400 "Bad request" error and not a 502 "Bad gateway" error, but I don't mind.

Reply:
I.e., I'm OK with leaving it kind of like "undefined behaviour"